battle.go

```diff
@@ -1,115 +1,212 @@
 package pokemonbattlelib
 
 import (
-	"log"
+	"fmt"
 )
 
 type Battle struct {
-	Weather  int  // one of the 6 in-battle weather conditions
-	ShiftSet bool // shift or set battle style for NPC trainer battles
-	State    BattleState
-	Players  []*Agent
+	Weather      int  // one of the 6 in-battle weather conditions
+	ShiftSet     bool // shift or set battle style for NPC trainer battles
+	State        BattleState
+	Parties      []*Party
+	Agents       []*Agent
+	agentParties map[int]int // Maps agent indexes to party indexes
+	teams        [][]int     // An array of teams, which are arrays of Agents. Used to derive allies and opponents
```

Comment:
The index of an Agent is it's ID.

The primary reason for agents to not manage their parties is so that they can't cheat by directly modifying the values of pokemon. That's why `battleContext` provides _copies_ of the data to the agent, and not pointers.

In a double battle, 2v1 trainers is a common scenario, where the lone trainer has a party, and the opponents have 2 different parties. This implementation covers this, others like it pretty nicely. It also allows us to expand pretty easily to doing triple battles, or battle royales, if we want to expand the scope to later pokemon generations.